### PR TITLE
[AIRFLOW-6857] Bulk sync DAGs

### PR DESCRIPTION
**Pull request by mik-laj**
I created the following DAG file:
```python
args = {
    'owner': 'airflow',
    'start_date': days_ago(3),
}


def create_dag(dag_number):
    dag = DAG(
        dag_id=f'perf_50_dag_dummy_tasks_{dag_number}_of_50', default_args=args,
        schedule_interval=None,
        dagrun_timeout=timedelta(minutes=60)
    )

    for j in range(1, 5):
        DummyOperator(
            task_id='task_{}_of_5'.format(j),
            dag=dag
        )

    return dag


for i in range(1, 200):
    globals()[f"dag_{i}"] = create_dag(i)

```
and I used the following code to test performance.
```python
import functools
import logging

import time

from airflow.jobs.scheduler_job import DagFileProcessor


class CountQueries(object):
    def __init__(self):
        self.count = 0

    def __enter__(self):
        from sqlalchemy import event
        from airflow.settings import engine
        event.listen(engine, "after_cursor_execute", self.after_cursor_execute)
        return None

    def after_cursor_execute(self, *args, **kwargs):
        self.count += 1

    def __exit__(self, type, value, traceback):
        from sqlalchemy import event
        from airflow.settings import engine
        event.remove(engine, "after_cursor_execute", self.after_cursor_execute)
        print('Query count: ', self.count)


count_queries = CountQueries

DAG_FILE = "/files/dags/200_dag_5_dummy_tasks.py"

log = logging.getLogger("airflow.processor")
processor = DagFileProcessor([], log)

def timing(f):

    @functools.wraps(f)
    def wrap(*args):
        RETRY_COUNT = 5
        r = []
        for i in range(RETRY_COUNT):
            time1 = time.time()
            f(*args)
            time2 = time.time()
            diff = (time2 - time1) * 1000.0
            r.append(diff)
            # print('Retry %d took %0.3f ms' % (i, diff))
        print('Average took %0.3f ms' % (sum(r) / RETRY_COUNT))

    return wrap


@timing
def slow_case():
    with count_queries():
        processor.process_file(DAG_FILE, None, pickle_dags=False)

slow_case()
```
I also cherry-picked AIRFLOW-6856

As a result, I obtained the following values

**Master**:
Query count: 1792
Average took 4505.891 ms

**AIRFLOW-6856:**
Query count: 1197
Average took 3203.710 ms

**Current:**
Query count: 602 
Average time: 2018.891 ms

**Diff to AIRFLOW-6856**
Query count: -592 (-49%)
Average time: -1185ms (-36%)

**Diff to master**
Query count: -1190 (-66%)
Average time: -2484 ms (-55%)

Thanks for support to @evgenyshulman from Databand!

---
Issue link: [AIRFLOW-6857](https://issues.apache.org/jira/browse/AIRFLOW-6857)

Make sure to mark the boxes below before creating PR: [x]

- [X] Description above provides context of the change
- [X] Commit message/PR title starts with `[AIRFLOW-NNNN]`. AIRFLOW-NNNN = JIRA ID<sup>*</sup>
- [X] Unit tests coverage for changes (not needed for documentation changes)
- [X] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [X] Relevant documentation is updated including usage instructions.
- [X] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

<sup>*</sup> For document-only changes commit message can start with `[AIRFLOW-XXXX]`.

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Review by turbaszek — APPROVED** [review]

File: airflow/models/dag.py [hunk 1]
```diff
@@ -1448,63 +1448,88 @@ def create_dagrun(self,
...
-        orm_dag.owners = owner
-        orm_dag.is_active = True
-        orm_dag.last_scheduler_run = sync_time
-        orm_dag.default_view = self._default_view
-        orm_dag.description = self.description
-        orm_dag.schedule_interval = self.schedule_interval
-        orm_dag.tags = self.get_dagtags(session=session)
+            orm_dags.append(orm_dag)
+
+        for orm_dag in orm_dags:
```
Comment (potiuk) [inline airflow/models/dag.py]: I love the change! 

I have only one suggestion here. I am not sure if there will be parallel executions of  bulk_sync_db() but even if there are none now there might be some in the future.

In such case it's an excellent practice to force the same sequence of updates on multiple entries by sorting the entries in the same order. This prevents the deadlock situation. When multiple entries are processed in random order, deadlock situation is very likely to happen. Sorting them in a consistent order makes deadlock impossible to occur. 

in this case I think simple sorting of the orm_dag by the dag_id should do the trick. Also if we have any other bulk updates to dagbag table we should apply the same principle.

File: airflow/jobs/scheduler_job.py [hunk 2]
```diff
@@ -785,8 +785,7 @@ def process_file(self, file_path, zombies, pickle_dags=False, session=None):
             return [], len(dagbag.import_errors)
 
         # Save individual DAGs in the ORM and update DagModel.last_scheduled_time
-        for dag in dagbag.dags.values():
-            dag.sync_to_db()
+        DAG.bulk_sync_to_db(dagbag.dags.values())
```
Comment (ashb) [inline airflow/jobs/scheduler_job.py]: I wonder if this should live on DagBag:

```suggestion
        dagbag.bulk_sync_to_db()
```

File: airflow/models/dag.py [hunk 3]
```diff
@@ -1448,63 +1448,88 @@ def create_dagrun(self,
...
-
-        orm_dag = session.query(
-            DagModel).filter(DagModel.dag_id == self.dag_id).first()
-        if not orm_dag:
-            orm_dag = DagModel(dag_id=self.dag_id)
-            if self.is_paused_upon_creation is not None:
-                orm_dag.is_paused = self.is_paused_upon_creation
-            self.log.info("Creating ORM DAG for %s", self.dag_id)
+        log.info("Sync %s DAGs", len(dags))
+        dag_by_ids = {dag.dag_id: dag for dag in dags}
```
Comment (ashb) [inline airflow/models/dag.py:1479]: If we made this a method on DagBag instead dag_by_ids is `self.dags` - it's possibly a micro-optimization, but right now the only case of "bulk_sync" is being called with `DAG.bulk_sync_to_db(dagbag.dags.values())` -- so we're throwing away the info we've already got to re-compute it.

File: airflow/models/dag.py (hunk 1 above)
Comment (ashb) [reply airflow/models/dag.py]: I can almost guarantee that in the future there _will_ become parallel executions of this method, my next 2 months of work is going in planning (working out an AIP now) to make Scheduler HA work and be officially supported.

Comment (ashb) [reply airflow/models/dag.py]: To that end, it is probably worth putting a `.with_for_update()` on to the query that produces orm_dags.

File: airflow/models/dag.py [hunk 4]
```diff
@@ -1448,63 +1448,88 @@ def create_dagrun(self,
...
+            else:
+                orm_dag.is_subdag = False
+                orm_dag.fileloc = dag.fileloc
+                orm_dag.owners = dag.owner
+            orm_dag.is_active = True
+            orm_dag.last_scheduler_run = sync_time
+            orm_dag.default_view = dag._default_view
+            orm_dag.description = dag.description
+            orm_dag.schedule_interval = dag.schedule_interval
+            orm_dag.tags = dag.get_dagtags(session=session)
```
Comment (ashb) [inline airflow/models/dag.py:1518]: This I think is going to go back to the DB, ignoring the joinedload you did earlier? Or does this somehow work because of the smarts in the SQLASession?

File: tests/models/test_dag.py [hunk 5]
```diff
@@ -739,13 +739,13 @@ def test_sync_to_db(self, mock_now):
         )
         with dag:
             DummyOperator(task_id='task', owner='owner1')
+            subdag = DAG('dag.subtask', start_date=DEFAULT_DATE, )
+            subdag.parent_dag = dag
+            subdag.is_subdag = True
```
Comment (ashb) [inline tests/models/test_dag.py:787]: Is this change needed? Doesn't the SubDagOperator set these automatically? If not it should -- I don't think users should have to set these properties to get "working" subdags.

**Review by ashb — CHANGES_REQUESTED** [review]
A nice change overall!

Couple of questions, and a possible different location for the bulk sync method.

File: tests/models/test_dag.py (hunk 5 above)
Comment (mik-laj) [reply tests/models/test_dag.py:787]: It was set by DagBag. We don't use DagBag, so this value is not set.
https://github.com/apache/airflow/blob/master/airflow/models/dagbag.py#L374-L378

Comment (ashb) [reply tests/models/test_dag.py:787]: Gotcha. Maybe worth a comment here just to explain why we need to set it in this case.

File: airflow/models/dag.py (hunk 4 above)
Comment (mik-laj) [reply airflow/models/dag.py:1518]: SQLAlchemy is smart enough.
When I did following changes:
```diff
-        orm_dags = session.query(DagModel) \
-            .options(
-            joinedload(DagModel.tags, innerjoin=False)
-        ) \
-            .filter(DagModel.dag_id.in_(dag_ids)).all()
+        orm_dags = session.query(DagModel)\
+            .filter(DagModel.dag_id.in_(dag_ids)).all()
``` 
I got following values - 
Before - 200 DAGs:
```
Query count:  602
Average took 3736.448 ms
```
After:
```
Query count:  801
Average took 3946.783 ms
```

Comment (ashb) [reply airflow/models/dag.py:1518]: ~I think you have before and after backwards here :)~

Ah no I see. 

Anyway, cool

File: tests/models/test_dag.py (hunk 5 above)
Comment (mik-laj) [reply tests/models/test_dag.py:787]: I added a comment.

File: airflow/models/dag.py (hunk 3 above)
Comment (mik-laj) [reply airflow/models/dag.py:1479]: I checked and it doesn't have a big impact on performance. I left the current code to keep the code legible.

File: airflow/jobs/scheduler_job.py (hunk 2 above)
Comment (mik-laj) [reply airflow/jobs/scheduler_job.py]: I created a helper method in DagBag.

File: airflow/models/dag.py (hunk 1 above)
Comment (mik-laj) [reply airflow/models/dag.py]: I added `sorted` and `with_for_update`.

Comment (potiuk) [reply airflow/models/dag.py]: Beautiful ! Indeed "for update" is a good idea as well!

**Comment by mik-laj** [comment]
I have following error, so I have to delete wait_for_update.
```
E       sqlalchemy.exc.NotSupportedError: (psycopg2.errors.FeatureNotSupported) FOR UPDATE cannot be applied to the nullable side of an outer join
E       
E       [SQL: SELECT dag.dag_id AS dag_dag_id, dag.root_dag_id AS dag_root_dag_id, dag.is_paused AS dag_is_paused, dag.is_subdag AS dag_is_subdag, dag.is_active AS dag_is_active, dag.last_scheduler_run AS dag_last_scheduler_run, dag.last_pickled AS dag_last_pickled, dag.last_expired AS dag_last_expired, dag.scheduler_lock AS dag_scheduler_lock, dag.pickle_id AS dag_pickle_id, dag.fileloc AS dag_fileloc, dag.owners AS dag_owners, dag.description AS dag_description, dag.default_view AS dag_default_view, dag.schedule_interval AS dag_schedule_interval, dag_tag_1.name AS dag_tag_1_name, dag_tag_1.dag_id AS dag_tag_1_dag_id 
E       FROM dag LEFT OUTER JOIN dag_tag AS dag_tag_1 ON dag.dag_id = dag_tag_1.dag_id 
E       WHERE dag.dag_id IN (%(dag_id_1)s, %(dag_id_2)s, %(dag_id_3)s, %(dag_id_4)s, %(dag_id_5)s, %(dag_id_6)s, %(dag_id_7)s, %(dag_id_8)s, %(dag_id_9)s, %(dag_id_10)s, %(dag_id_11)s, %(dag_id_12)s, %(dag_id_13)s, %(dag_id_14)s, %(dag_id_15)s, %(dag_id_16)s, %(dag_id_17)s, %(dag_id_18)s, %(dag_id_19)s, %(dag_id_20)s, %(dag_id_21)s, %(dag_id_22)s, %(dag_id_23)s, %(dag_id_24)s, %(dag_id_25)s, %(dag_id_26)s, %(dag_id_27)s, %(dag_id_28)s, %(dag_id_29)s, %(dag_id_30)s, %(dag_id_31)s, %(dag_id_32)s, %(dag_id_33)s, %(dag_id_34)s, %(dag_id_35)s, %(dag_id_36)s, %(dag_id_37)s, %(dag_id_38)s, %(dag_id_39)s, %(dag_id_40)s, %(dag_id_41)s, %(dag_id_42)s, %(dag_id_43)s, %(dag_id_44)s, %(dag_id_45)s, %(dag_id_46)s, %(dag_id_47)s, %(dag_id_48)s, %(dag_id_49)s, %(dag_id_50)s, %(dag_id_51)s, %(dag_id_52)s) FOR UPDATE]
E       [parameters: {'dag_id_1': 'test_on_kill', 'dag_id_2': 'example_external_task_marker_child', 'dag_id_3': 'test_utils', 'dag_id_4': 'example_xcom', 'dag_id_5': 'test_mark_success', 'dag_id_6': 'test_localtaskjob_double_trigger', 'dag_id_7': 'test_dagrun_states_deadlock', 'dag_id_8': 'example_branch_dop_operator_v3', 'dag_id_9': 'test_depends_on_past', 'dag_id_10': 'test_dagrun_states_fail', 'dag_id_11': 'impersonation_subdag.test_subdag_operation', 'dag_id_12': 'test_dagrun_states_root_fail', 'dag_id_13': 'clear_subdag_test_dag.daily_job', 'dag_id_14': 'test_dagrun_states_root_fail_unfinished', 'dag_id_15': 'example_passing_params_via_test_command', 'dag_id_16': 'latest_only_with_trigger', 'dag_id_17': 'example_subdag_operator.section-1', 'dag_id_18': 'test_start_date_scheduling', 'dag_id_19': 'test_backfill_pooled_task_dag', 'dag_id_20': 'latest_only', 'dag_id_21': 'test_dagrun_states_root_future', 'dag_id_22': 'test_task_start_date_scheduling', 'dag_id_23': 'example_kubernetes_executor', 'dag_id_24': 'test_latest_runs_1', 'dag_id_25': 'test_task_view_type_check', 'dag_id_26': 'example_trigger_target_dag', 'dag_id_27': 'impersonation_subdag', 'dag_id_28': 'test_default_impersonation', 'dag_id_29': 'test_dag_under_subdir2', 'dag_id_30': 'test_zip_dag', 'dag_id_31': 'test_retry_handling_job', 'dag_id_32': 'test_no_impersonation', 'dag_id_33': 'example_short_circuit_operator', 'dag_id_34': 'example_bash_operator', 'dag_id_35': 'test_run_ignores_all_dependencies', 'dag_id_36': 'test_subdag_operator', 'dag_id_37': 'example_subdag_operator.section-2', 'dag_id_38': 'example_kubernetes_executor_config', 'dag_id_39': 'example_branch_operator', 'dag_id_40': 'test_example_bash_operator', 'dag_id_41': 'example_trigger_controller_dag', 'dag_id_42': 'example_complex', 'dag_id_43': 'example_external_task_marker_parent', 'dag_id_44': 'test_subdag_operator.section-1', 'dag_id_45': 'tutorial', 'dag_id_46': 'example_subdag_operator', 'dag_id_47': 'test_heartbeat_failed_fast', 'dag_id_48': 'example_python_operator', 'dag_id_49': 'example_skip_dag', 'dag_id_50': 'test_impersonation', 'dag_id_51': 'clear_subdag_test_dag', 'dag_id_52': 'test_dagrun_states_success'}]
```

**Review by ashb — APPROVED** [review]

**Comment by ashb** [comment]
The `FOR UPDATE` is a fairly hard requirement. Without it scheduler HA is impossible to do correctly. :(

**Comment by ashb** [comment]
Ah, cos Postgres is trying to lock everything.

@mik-laj `.with_for_update(of=DagModel)` should be what we need.

**Review by houqp — APPROVED** [review]

File: airflow/models/dag.py [hunk 6]
```diff
@@ -1455,63 +1455,92 @@ def create_dagrun(self,
...
 
         session.commit()
```
Comment (saguziel) [inline airflow/models/dag.py:1520]: it might be worth doing this in smaller batches rather than one gigantic commit. I don't think atomicity is a concern but large transactions may be.

Comment (mik-laj) [reply airflow/models/dag.py:1520]: This is only done from DAGs from one file. I suspect that even if someone puts 200+ DAG in one file, adding up to 200+ entry to DB in one transaction is not a problem.

Comment (mik-laj) [reply airflow/models/dag.py]: Thanks. It will improve soon.

**Review by Fokko — APPROVED** [review]
LGTM

**Review by feluelle — APPROVED** [review]
Nice one! 👍
